orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,733 @@
+        }
+        else if (op == DEL_COMMAND)
+        {
+            deleteEntry(key);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+        }
+
+        consumer.m_toSync.erase(it++);
+    }
+}
+
+void PfcWdOrch::updateWdCounters(const std::string& queueIdStr, bool operational)
+{
+    uint32_t detectCount = 0;
+    uint32_t restoreCount = 0;
+    std::vector<FieldValueTuple> resultFvValues;
+
+    getWdCounters(queueIdStr, detectCount, restoreCount);
+
+    if (operational)
+    {
+        restoreCount++;
+    }
+    else
+    {
+        detectCount++;
+    }
+
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED, std::to_string(detectCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED, std::to_string(restoreCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATUS, operational ?
+                                                     PFC_WD_QUEUE_STATUS_OPERATIONAL :
+                                                     PFC_WD_QUEUE_STATUS_STORMED);
+
+    m_countersTable.set(queueIdStr, resultFvValues);
+}
+
+void PfcWdOrch::initWdCounters(const std::string &queueIdStr)
+{
+    uint32_t detectCount = 0;
+    uint32_t restoreCount = 0;
+    std::vector<FieldValueTuple> resultFvValues;
+
+    getWdCounters(queueIdStr, detectCount, restoreCount);
+
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED, std::to_string(detectCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED, std::to_string(restoreCount));
+    resultFvValues.emplace_back(PFC_WD_QUEUE_STATUS, PFC_WD_QUEUE_STATUS_OPERATIONAL);
+
+    m_countersTable.set(queueIdStr, resultFvValues);
+}
+
+void PfcWdOrch::getWdCounters(const std::string& queueIdStr, uint32_t& detectCount, uint32_t& restoreCount)
+{
+    std::vector<FieldValueTuple> fieldValues;
+    detectCount = 0;
+    restoreCount = 0;
+
+    if (!m_countersTable.get(queueIdStr, fieldValues))
+    {
+        return;
+    }
+
+    for (const auto& fv : fieldValues)
+    {
+        const auto field = fvField(fv);
+        const auto value = fvValue(fv);
+
+        if (field == PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED)
+        {
+            detectCount = std::stoul(value);
+        }
+        else if (field == PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED)
+        {
+            restoreCount = std::stoul(value);
+        }
+    }
+}
+
+PfcWdOrch::PfcWdAction PfcWdOrch::deserializeAction(const string& key)
+{
+    SWSS_LOG_ENTER();
+
+    const std::map<std::string, PfcWdAction> actionMap =
+    {
+        { "forward", PfcWdAction::PFC_WD_ACTION_FORWARD },
+        { "drop", PfcWdAction::PFC_WD_ACTION_DROP },
+    };
+
+    if (actionMap.find(key) == actionMap.end())
+    {
+        return PfcWdAction::PFC_WD_ACTION_UNKNOWN;
+    }
+
+    return actionMap.at(key);
+}
+
+void PfcWdOrch::createEntry(const string& key, const vector<FieldValueTuple>& data)
```

Comment:
CreateEntry looks too big for me. I would propose define StartWdOnPort, which includes code starting from line 250, and rename StartWd to StartWdOnQueue. We could make StartWdonPort and StartOnQueue virtual function. Another reason is that we need to set sai api for other platforms which is slight different from the current implementation.  Overriding the whole createEntry function will have quite a lot duplicate code. If we separate the StartWdOnPort logic out, it will be much easier to support all platforms.

Reply:
Done. The common interface has StartWdOnPort.